rosbag2_transport/src/rosbag2_transport/topic_filter.cpp

```diff
@@ -94,21 +94,31 @@ TopicFilter::TopicFilter(
   RecordOptions record_options,
   rclcpp::node_interfaces::NodeGraphInterface::SharedPtr node_graph,
   bool allow_unknown_types)
-: record_options_(record_options),
+: record_options_(std::move(record_options)),
   allow_unknown_types_(allow_unknown_types),
   node_graph_(node_graph)
 {}
 
-TopicFilter::~TopicFilter()
-{}
+TopicFilter::~TopicFilter() = default;
 
 std::unordered_map<std::string, std::string> TopicFilter::filter_topics(
   const std::map<std::string, std::vector<std::string>> & topic_names_and_types)
 {
   std::unordered_map<std::string, std::string> filtered_topics;
   for (const auto & [topic_name, topic_types] : topic_names_and_types) {
-    if (take_topic(topic_name, topic_types)) {
-      filtered_topics.insert(std::make_pair(topic_name, topic_types[0]));
+    // Check take_topics_cache_ first to avoid performance burden when discovery thread running
+    auto take_topics_cache_it = take_topics_cache_.find(topic_name);
+    if (take_topics_cache_it != take_topics_cache_.end()) {
+      if (take_topics_cache_it->second) {
+        filtered_topics.insert(std::make_pair(topic_name, topic_types[0]));
+      }
+    } else {
+      if (take_topic(topic_name, topic_types)) {
+        filtered_topics.insert(std::make_pair(topic_name, topic_types[0]));
+        take_topics_cache_[topic_name] = true;
+      } else {
+        take_topics_cache_[topic_name] = false;
```

Comment:
1st discovery, lets say if the topic is bound with unknown type, and the register here with `false`. after 2nd discovery, even if the topic is rebound to the appropriate type, it cannot be taken since it does not reconsider the topic type every time. how about using `already_warned_unknown_types_` to check if we already know that is unknown type or not to deny the topic instead?